include/gatekeeper_ipip.h

```diff
@@ -32,10 +33,8 @@
 #define IP_DN_FRAGMENT_FLAG     (0x0040)
 
 struct ipip_tunnel_info {
-	struct ip_flow	     flow;
-	struct ether_addr    source_mac;
-	/* TODO The MAC addresses must come from the LLS block. */
-	struct ether_addr    nexthop_mac;
+	struct ip_flow	 flow;
+	struct ether_hdr *eth_hdr;
 };
```

Comment:
This struct combining IP and Ethernet information is just creating trouble, drop it and use `struct ip_flow` and `struct ether_hdr` where needed. You should see some of the code to straighten up. If this is still not clear the why, try to keep it the way it's now and fix the policy table as the other comments are suggesting.